src/os/bluestore/BlueStore.cc

```diff
@@ -3675,10 +3686,25 @@ void BlueStore::ExtentMap::reshard(
     }
     dout(30) << " extent " << *e << dendl;
 
-    // disfavor shard boundaries that span a blob
-    bool would_span = (e->logical_offset < max_blob_end) || e->blob_offset;
-    if (estimate &&
-	estimate + extent_avg > target + (would_span ? slop : 0)) {
+    bool make_shard_here = false;
+    if (onode_data_has_boundaries) {
+      if (e->blob_start() >= next_boundary) {
+	// this it the place we want to have shard boundary
+	if ((estimate >= target /*we have enough already*/) ||
+	    (estimate + encoded_segment_estimate >= target * 3 / 2)
+	    /*we will be too large if we wait for next segment*/) {
+	  make_shard_here = true;
+	}
+	next_boundary = p2align(e->blob_end() + data_segment_size, data_segment_size);
```

Comment:
what if blob_end() points exactly to the next segment (e.g. blob_start = 0 and blob_end() = 1M and segment size = 1M) - then next boundary to be equal to 2MB which is apparently not what we expect (1M)

Reply:
Fixed. Changed to p2roundup.